core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java

```diff
@@ -0,0 +1,336 @@
+    }
+
+    @Test
+    public void testDelegateToMaster() throws ExecutionException, InterruptedException {
+        Request request = new Request().masterNodeTimeout(TimeValue.timeValueSeconds(3));
+        clusterService.setState(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_NOT_MASTER));
+
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        final AtomicBoolean delegationToMaster = new AtomicBoolean();
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected void processBeforeDelegationToMaster(Request request, ClusterState state) {
+                logger.debug("Delegation to master called");
+                delegationToMaster.set(true);
+            }
+        }.execute(request, listener);
+
+        assertTrue("processBeforeDelegationToMaster not called", delegationToMaster.get());
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        CapturingTransport.CapturedRequest capturedRequest = transport.capturedRequests()[0];
+        assertTrue(capturedRequest.node.isMasterNode());
+        assertThat(capturedRequest.request, equalTo(request));
+        assertThat(capturedRequest.action, equalTo("testAction"));
+
+        Response response = new Response();
+        transport.handleResponse(capturedRequest.requestId, response);
+        assertThat(listener.get(), equalTo(response));
+    }
+
+    @Test
+    public void testDelegateToFailingMaster() throws ExecutionException, InterruptedException {
+        boolean fallbackToLocalBeforeTimeout = randomBoolean();
+        Request request = new Request().masterNodeTimeout(TimeValue.timeValueSeconds(fallbackToLocalBeforeTimeout ? 60 : 1));
+        clusterService.setState(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_NOT_MASTER));
+
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        final AtomicBoolean delegationToMaster = new AtomicBoolean();
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected void processBeforeDelegationToMaster(Request request, ClusterState state) {
+                logger.debug("Delegation to master called");
+                delegationToMaster.set(true);
+            }
+        }.execute(request, listener);
+
+        assertTrue("processBeforeDelegationToMaster not called", delegationToMaster.get());
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        CapturingTransport.CapturedRequest capturedRequest = transport.capturedRequests()[0];
+        assertTrue(capturedRequest.node.isMasterNode());
+        assertThat(capturedRequest.request, equalTo(request));
+        assertThat(capturedRequest.action, equalTo("testAction"));
+
+        transport.handleResponse(capturedRequest.requestId, new ConnectTransportException(remoteNode, "Fake network error"));
```

Comment:
can we also have a test with some exception other than connect and check that it returns immediately?